contributing/building-a-bento-amp-extension.md

```diff
@@ -0,0 +1,681 @@
+
+Before diving into the details on creating a new Bento AMP extension, please familiarize yourself with the [general process for contributing code and features to AMP](https://github.com/ampproject/amphtml/blob/master/contributing/contributing-code.md). Since you are adding a new extension you will likely need to follow the [process for making a significant change](https://github.com/ampproject/amphtml/blob/master/contributing/contributing-code.md#process-for-significant-changes), including filing an ["Intent to Implement" issue](https://github.com/ampproject/amphtml/labels/INTENT%20TO%20IMPLEMENT) and finding a guide before you start significant development.
+
+## Naming
+
+All Bento AMP extensions have their tag names prefixed with `amp-`.
+Make sure to choose an accurate and clear name for your extension.
+
+Extensions that embed a third-party service must follow the [guidelines for naming a third-party component](../spec/amp-3p-naming.md).
+
+## Directory structure
+
+You create your extension's files inside the `extensions/` directory.
+The directory structure is below:
+
+```sh
+/extensions/amp-my-element/
+├── 0.1/
+|   ├── storybook/                       # Element's manual sample playground (req'd)
+|   |   ├── Basic.js                     # Preact usage examples
+|   |   ├── Basic.amp.js                 # AMP usage examples
+|   ├── test/                            # Element's unit test suite (req'd)
+|   |   ├── test-amp-my-element.js
+|   |   └── More test JS files (optional)
+|   ├── my-element.js                    # Preact implementation (req'd)
+|   ├── my-element.type.css              # Preact type definitions (req'd)
+|   ├── amp-my-element.js                # Element's implementation (req'd)
+|   ├── amp-my-element.css               # Custom CSS (optional)
+|   ├── my-element.jss.js                # Preact JSS (optional)
+|   └── More JS files (optional)
+├── validator-amp-my-element.protoascii  # Validator rules (req'd)
+├── amp-my-element.md                    # Element's main documentation (req'd)
+└── More documentation in .md files (optional)
+└── OWNERS                               # Owners file. Primary contact(s) for the extension. (req'd)
+```
+
+In most cases you'll only create the required (req'd) files. If your element does not need custom CSS or JSS, you don't need to create those files.
+
+## Extend AMP.PreactBaseElement
+
+Almost all AMP extensions extend AMP.BaseElement, which provides some
+hookups and callbacks for you to override in order to implement and
+customize your element behavior. These callbacks are explained in
+the BaseElement Callbacks section of [Building an AMP Extension](https://github.com/ampproject/amphtml/blob/master/contributing/building-an-amp-extension.md), and are also explained inline in the
+[BaseElement](https://github.com/ampproject/amphtml/blob/master/src/base-element.js)
+class.
+
+All Bento AMP extensions extend AMP.PreactBaseElement, which builds upon
```

Comment:
Probably "All Preact-based Bento AMP ..." since above we say that "Preact" is optional.